File: solutions/_103_binary_tree_zigzag_level_order_traversal/Solution.java
package _103_binary_tree_zigzag_level_order_traversal;

import java.util.*;

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null)
            return res;

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean reverse = false;

        while (!q.isEmpty()) {
            int size = q.size();
            List<Integer> level = new ArrayList<>();

            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                if (reverse) {
                    level.add(0, node.val);
                } else {
                    level.add(node.val);
                }

                if (node.left != null)
                    q.add(node.left);
                if (node.right != null)
                    q.add(node.right);
            }

            res.add(level);
            reverse = !reverse;
        }

        return res;

    }
}

/*
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 */

File: solutions/_103_binary_tree_zigzag_level_order_traversal/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
  if (!root) {
    return [];
  }

  const result = [];
  const q = [root];
  let reverse = false;

  while (q.length > 0) {
    const level = [];
    const n = q.length;

    for (let i = 0; i < n; i++) {
      const node = q.shift();
      if (reverse) {
        level.unshift(node.val);
      } else {
        level.push(node.val);
      }

      if (node.left) {
        q.push(node.left);
      }
      if (node.right) {
        q.push(node.right);
      }
    }

    result.push(level);
    reverse = !reverse;
  }

  return result;
};

module.exports = { zigzagLevelOrder, TreeNode };

/*
https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
*/
